spec/app/ics-020-fungible-token-transfer/README.md

```diff
@@ -210,33 +253,54 @@ In plain English, between chains `A` and `B`:
   an acknowledgement of failure is preferable to aborting the transaction since it more easily enables the sending chain
   to take appropriate action based on the nature of the failure.
 
+Note: `constructOnChainDenom` is a helper function that will construct the local on-chain denomination for the bridged token. It **must** encode the trace and base denomination to ensure that tokens coming over different paths are not treated as fungible. The original trace and denomination must be retrievable by the state machine so that they can be passed in their original forms when constructing a new IBC path for the bridged token. The ibc-go implementation handles this by creating a local denomination: `hash(trace+base_denom)`.
+
 `sendFungibleTokens` must be called by a transaction handler in the module which performs appropriate signature checks, specific to the account owner on the host state machine.
 
 ```typescript
 function sendFungibleTokens(
-  denomination: string,
-  amount: uint256,
+  tokens: []Token,
   sender: string,
   receiver: string,
+  memo: string,
   sourcePort: string,
   sourceChannel: string,
   timeoutHeight: Height,
   timeoutTimestamp: uint64): uint64 {
-    prefix = "{sourcePort}/{sourceChannel}/"
-    // we are the source if the denomination is not prefixed
-    source = denomination.slice(0, len(prefix)) !== prefix
-    if source {
-      // determine escrow account
-      escrowAccount = channelEscrowAddresses[sourceChannel]
-      // escrow source tokens (assumed to fail if balance insufficient)
-      bank.TransferCoins(sender, escrowAccount, denomination, amount)
-    } else {
-      // receiver is source chain, burn vouchers
-      bank.BurnCoins(sender, denomination, amount)
+    for token in tokens {
+      prefix = "{sourcePort}/{sourceChannel}/"
+      // we are the source if the denomination is not prefixed
+      source = token.trace[0] !== prefix
+      onChainDenom = constructOnChainDenom(token.trace, token.denom)
+      
+      if source {
+        // determine escrow account
+        escrowAccount = channelEscrowAddresses[sourceChannel]
+        // escrow source tokens (assumed to fail if balance insufficient)
+        bank.TransferCoins(sender, escrowAccount, onChainDenom, token.amount)
+      } else {
+        // receiver is source chain, burn vouchers
+        bank.BurnCoins(sender, onChainDenom, token.amount)
+      }
     }
 
-    // create FungibleTokenPacket data
-    data = FungibleTokenPacketData{denomination, amount, sender, receiver}
+    channel = provableStore.get(channelPath(sourcePort, sourceChannel))
+    if channel.version === "ics20-1" {
+      abortTransactionUnless(len(tokens) == 1)
+      data = FungibleTokenPacketData{tokens[0].denom, tokens[0].amount, sender, receiver, memo}
```

Comment:
I don't think this is correct, 

`tokens[0].denom` is not the same as `FungibleTokenPacketData.Denom`, the former is something like `atom` while the latter is `transfer/channel-0/transfer/channel-1/atom`.

Reply:
It might be good to be explicit and mention as well somewhere in the spec (maybe where the `Token` type is introduced) that `token.denom` is the base denomination, since in ibc-go [the denomination in the `sdk.Coins` that are passed to `sendTransfer`](https://github.com/cosmos/ibc-go/blob/749caee7f7a8ca902e94167d630fabe61d3e1e67/modules/apps/transfer/keeper/relay.go#L60) might be the IBC denomination ([from which we derive the trace and base denom information](https://github.com/cosmos/ibc-go/blob/749caee7f7a8ca902e94167d630fabe61d3e1e67/modules/apps/transfer/keeper/relay.go#L98)). Just to avoid confusion if someone reads the spec and compares with the code.

Reply:
I address this [here](https://github.com/cosmos/ibc/pull/1020/commits/d49a90bbb6ed28c7fcda99cb27ef209886ebad0b).